proptest-derive/src/ast.rs

```diff
@@ -0,0 +1,712 @@
+            ToReg::API => call_site_ident(API_PARAM_NAME).to_tokens(tokens),
+        }
+    }
+}
+
+impl ToTokens for Ctor {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        // The logic of each of these are pretty straight forward save for
+        // union which is described separately.
+        use self::Ctor::*;
+        match self {
+            Filter(ctor, filter) => quote_append!(tokens,
+                _proptest::strategy::Strategy::prop_filter(
+                    #ctor, stringify!(#filter), #filter)
+            ),
+            Extract(ctor, to, from) => quote_append!(tokens, {
+                let #to = #from; #ctor
+            }),
+            Arbitrary(ty, fv, span) => tokens.append_all(if let Some(fv) = fv {
+                let args = param(*fv);
+                quote_spanned!(*span=>
+                    _proptest::arbitrary::any_with::<#ty>(#args)
+                )
+            } else {
+                quote_spanned!(*span=>
+                    _proptest::arbitrary::any::<#ty>()
+                )
+            }),
+            Existential(expr) => quote_append!(tokens,
+                _proptest::strategy::Strategy::boxed( #expr ) ),
+            Value(expr) => quote_append!(tokens, || #expr ),
+            ValueExistential(expr) => quote_append!(tokens,
+                _proptest::strategy::Strategy::boxed(
+                    _proptest::strategy::LazyJust::new(move || #expr)
+                )
+            ),
+            Map(ctors, closure) => {
+                let ctors = ctors.iter();
+                quote_append!(tokens,
+                    _proptest::strategy::Strategy::prop_map(
+                        ( #(#ctors,)* ),
+                        #closure
+                    )
+                );
+            },
+            Union(ctors) => union_ctor_to_tokens(tokens, ctors),
+        }
+    }
+}
+
+/// Tokenizes a weighted list of `Ctor`.
+///
+/// The logic is that the output should be as linear as possible while still
+/// supporting enums with an unbounded number of variants without any boxing
+/// (erasure) or dynamic dispatch.
+///
+/// As `TupleUnion` is (currently) limited to 10 summands in the coproduct we
+/// can't just emit the entire thing linearly as this will fail on the 11:th
+/// variant.
+///
+/// A naive approach to solve might be to simply use a cons-list like so:
+///
+/// ```ignore
+/// TupleUnion::new(
+///     (w_1, s_1),
+///     (w_2 + w_3 + w_4 + w_5,
+///      TupleUnion::new(
+///         (w_2, s_2),
+///         (w_3 + w_4 + w_5,
+///          TupleUnion::new(
+///             (w_3, s_3),
+///             (w_4 + w_5,
+///              TupleUnion::new(
+///                 (w_4, s_4),
+///                 (w_5, s_5),
+///             ))
+///         ))
+///     ))
+/// )
+/// ```
+///
+/// However, we can do better by being linear for the `10 - 1` first
+/// strategies and then switch to nesting like so:
+///
+/// ```ignore
+/// (1, 2, 3, 4, 5, 6, 7, 8, 9,
```

Comment:
This nesting isn't going to produce a uniform distribution. This is why `prop_oneof!` falls back to boxing everything so it can use a normal `Union` if it's over the limit.

Reply:
Oh, I see you're doing something with weights below.

Reply:
Yeah, the idea is that:
```rust
prop_oneof![ w1 => e1, w2 => e2, w3 => e3 ]
```
can be rewritten in terms of:
```rust
prop_oneof![ w1 => e1, (w2 + w3) => prop_oneof![ w2 => e2, w3 => e3 ] ]
```
and it commutes, so you can also say:
```rust
prop_oneof![ (w1 + w2) => prop_oneof![ w1 => e1, w2 => e2 ], w3 => e3 ]
```

This is then applied for larger groupings (the max number that `TupleUnion` supports - 1).

In practice this would differ from the distribution you get when using `Union` + boxing because you are sampling the PRNG more times; but it seems to be that the cost of boxing every part of the union outweighs this slight discrepancy so I'd like to avoid the boxing.